File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        #Binary Tree so will have to explore both directions 
        #Keep track of the max value in each recursion to determine if add to counter
        counter = 0
        def dfs(node, val):
            nonlocal counter
            if not node: 
                return 
            elif node.val >= val:
                counter += 1
                dfs(node.left, node.val)
                dfs(node.right, node.val)
            else:
                dfs(node.left, val)
                dfs(node.right, val)
        dfs(root, root.val)
        return counter

File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {  
    public boolean isSubtree(TreeNode root, TreeNode subRoot) {
        if (subRoot == null){
            // Root tree always has a null value
            return true; 
        }
        if (root == null){
            //If the root is null and subroot is not, no match
            return false;
        }
        if(isTreeSub(root, subRoot)){
            return true;
        }
        return isSubtree(root.left, subRoot) ||
         isSubtree(root.right, subRoot);

        

    }
    public boolean isTreeSub(TreeNode root, TreeNode subRoot){
        if (root == null && subRoot == null){
            return true;
        }
        if (root != null && subRoot != null && root.val == subRoot.val){
            return isTreeSub(root.left, subRoot.left) && 
            isTreeSub(root.right, subRoot.right);
        }
        return false;
    }


}
